src/main/java/com/salesforce/dynamodbv2/mt/util/StreamsCache.java

```diff
@@ -0,0 +1,398 @@
+            }
+        }
+
+        /**
+         * Internal helper method to efficiently find the index of the given sequence number in the list of records.
+         *
+         * @param sequenceNumber Sequence number to find index of.
+         * @return Index in the list, such that all records in the list after the index have sequence numbers that are
+         *     greater or equal to the given sequence number. If no such records exist, returns the size of the list.
+         */
+        private int getIndex(BigInteger sequenceNumber) {
+            final List<BigInteger> sequenceNumbers = Lists.transform(records, ShardIteratorPosition::at);
+            int index = Collections.binarySearch(sequenceNumbers, sequenceNumber);
+            if (index < 0) {
+                index = (-index) - 1;
+            }
+            return index;
+        }
+
+        /**
+         * Returns the byte size (in the stream) of all records in this segment.
+         *
+         * @return Byte size of all records in this segment.
+         */
+        long getByteSize() {
+            return byteSize;
+        }
+
+        /**
+         * Returns whether this segment is empty. Note that non-empty segments may still have an empty records
+         * collections if the corresponding segment in the underlying stream contains no records for the sequence number
+         * range.
+         *
+         * @return True if this segment is empty, i.e., if {@link #start} is equal to {@link #end}. False, otherwise.
+         */
+        boolean isEmpty() {
+            return start.equals(end);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Segment segment = (Segment) o;
+            return start.equals(segment.start) && end.equals(segment.end) && records.equals(segment.records);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(start, end, records);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                .add("start", start)
+                .add("end", end)
+                .add("records", records)
+                .add("byteSize", byteSize)
+                .toString();
+        }
+    }
+
+    // config parameter
+    private final long maxRecordsByteSize;
+
+    // cached record segments sorted by sequence number within each shard
+    private final ConcurrentMap<ShardId, NavigableMap<BigInteger, Segment>> segments;
+    // Insertion order of cache segments for eviction purposes
+    private final Queue<ShardIteratorPosition> insertionOrder;
+    // locks for accessing shard caches
+    private final Striped<ReadWriteLock> shardLocks;
+    // size of cache >= 0
+    private final AtomicLong recordsByteSize;
```

Comment:
@msgroi @shirdoo FYI, based on your feedback, I have made the cache lock more fine-grained (at the granularity of a shard). Would be good to have you sanity-check that the desired invariants still hold under concurrent access. The main difference compared to the previous version (which was using a single lock across all shards) is that eviction is done after the lock is released.